File: ProjectStart/ProjectStart.WebApp/ClientApp/Shared/AppModels/NodeEntity.ts
import { BaseEntity, IBaseEntity} from 'shared/AppModels/BaseEntity';
/**
 * Interface for: ProjectStart.Entity.Common.NodeEntity<T>
 */
export interface INodeEntity<T> extends IBaseEntity {
    parentId: string;
    parentNode: T;
    subNodes: T[];
}
/**
 * Base view model for ProjectStart.Entity.Common.NodeEntity<T>
 */
export class NodeEntity<T> extends BaseEntity implements INodeEntity<T> {
    parentId: string;
    parentNode: T;
    subNodes: T[];
}

File: ProjectStart/ProjectStart.Entity/Common/NodeEntity.cs
using System.Collections.Generic;
using System.ComponentModel.DataAnnotations.Schema;

namespace ProjectStart.Entity.Common
{
    public class NodeEntity<T> : BaseEntity
    {
        [ForeignKey("ParentNode")]
        public virtual string ParentId { get; set; }
        public virtual T ParentNode { get; set; }
        public virtual ICollection<T> SubNodes { get; set; } = new HashSet<T>();
    }
}
